### PR TITLE
Remove `constant_function` for `AutoEnzyme`

### PR DESCRIPTION
**Pull request by gdalle**
> [!WARNING]
> This PR and the associated release are breaking wrt 1.6.1, but we will yank 1.6.1 and 1.6.0 right after

## Checklist

- [x] Appropriate tests were added
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API
  
## Additional context

Following a long discussion with @ChrisRackauckas and @wsmoses, we decided to remove the `constant_function` setting for now, and reinsert it later with different defaults (once Enzyme implements a few changes on their side). This PR does the removal and bumps ADTypes to v1.6.2, after which we will need to yank v1.6.0 and v1.6.1.

- https://github.com/EnzymeAD/Enzyme.jl/issues/1669

### CODE REVIEW

**Comment by gdalle** [comment]
Let's wait for a review from Billy and Chris before merging

**Comment by ChrisRackauckas** [comment]
Let's just go straight to the `nothing` default then?

**Comment by gdalle** [comment]
I wanted to wait until the implementation exists on the Enzyme and DI side before reintroducing this kwarg. Our mistake last time was to introduce it before it actually did anything, and as a result we got the design wrong.

**Comment by gdalle** [comment]
But if @wsmoses is confident that `constant_function=nothing` is the right third option I can modify this PR to just change the default instead of removing the kwarg

**Comment by wsmoses** [comment]
I would probably do the yank now then design when ready.

Being even more forward looking this should likely actually be a Union{Nothing, Enzyme.Annotation}, since the function could be nothing, constant, duplicated, or even active

**Comment by wsmoses** [comment]
This constant function is not the appropriate name

**Comment by ChrisRackauckas** [comment]
function_activity

**Comment by wsmoses** [comment]
I mean that sounds like a good name to me, but it also might be wise to
wait til things are settled before choosing

**Comment by gdalle** [comment]
Okay then removing everything first is the right move. In this spirit, do you approve this PR?

**Review by wsmoses — APPROVED** [review]
Spiritually and actually approved

**Pull request merged**
